internal/controller/driver_controller.go

```diff
@@ -18,45 +18,1120 @@ package controller
+
+	r.images = maps.Clone(imageDefaults)
+	if r.driver.Spec.ImageSet != nil {
+		imageSetCM := corev1.ConfigMap{}
+		imageSetCM.Name = r.driver.Spec.ImageSet.Name
+		imageSetCM.Namespace = r.driver.Namespace
+		if err := r.Get(r.ctx, client.ObjectKeyFromObject(&imageSetCM), &imageSetCM); err != nil {
+			r.log.Error(err, "Unable to load driver specified image set config map", "name", client.ObjectKeyFromObject(&imageSetCM))
+			return err
+		}
+
+		maps.Copy(r.images, imageSetCM.Data)
+	}
+
+	return nil
+}
+
+func (r *driverReconcile) reconcileK8sCsiDriver() error {
+	existingCsiDriver := &storagev1.CSIDriver{}
+	existingCsiDriver.Name = r.driver.Name
+
+	log := r.log.WithValues("driverName", existingCsiDriver.Name)
+	log.Info("Reconciling CSI Driver")
+
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(existingCsiDriver), existingCsiDriver); client.IgnoreNotFound(err) != nil {
+		log.Error(err, "Failed to load CSI Driver resource")
+		return err
+	}
+
+	desiredCsiDriver := existingCsiDriver.DeepCopy()
+	desiredCsiDriver.Spec = storagev1.CSIDriverSpec{
+		PodInfoOnMount: ptr.To(false),
+		AttachRequired: utils.If(
+			r.driver.Spec.AttachRequired != nil,
+			r.driver.Spec.AttachRequired,
+			ptr.To(true),
+		),
+		FSGroupPolicy: utils.If(
+			r.driver.Spec.FsGroupPolicy != "",
+			ptr.To(r.driver.Spec.FsGroupPolicy),
+			ptr.To(storagev1.FileFSGroupPolicy),
+		),
+	}
+
+	ownerObjKey := client.ObjectKeyFromObject(&r.driver)
+	if bytes, err := json.Marshal(ownerObjKey); err != nil {
+		log.Error(
+			err,
+			"Failed to JSON marshal owner obj key for CSI driver resource",
+			"ownerObjKey",
+			ownerObjKey,
+		)
+		return err
+	} else {
+		utils.AddAnnotation(desiredCsiDriver, ownerRefAnnotationKey, string(bytes))
+	}
+
+	if existingCsiDriver.UID == "" || !reflect.DeepEqual(desiredCsiDriver, existingCsiDriver) {
+		if existingCsiDriver.UID != "" {
+			log.Info("CSI Driver resource exist but does not meet desired state")
+			if err := r.Delete(r.ctx, existingCsiDriver); err != nil {
+				log.Error(err, "Failed to delete existing CSI Driver resource")
+				return err
+			}
+			log.Info("CSI Driver resource deleted successfully")
+		} else {
+			log.Info("CSI Driver resource does not exist")
+		}
+
+		if err := r.Create(r.ctx, desiredCsiDriver); err != nil {
+			log.Error(err, "Failed to create a CSI Driver resource")
+			return err
+		}
+
+		log.Info("CSI Driver resource created successfully")
+	} else {
+		log.Info("CSI Driver resource already meets desired state")
+	}
+
+	return nil
+}
+
+func (r *driverReconcile) reconcileControllerPluginDeployment() error {
+	deploy := &appsv1.Deployment{}
+	deploy.Name = r.generateName("ctrlplugin")
+	deploy.Namespace = r.driver.Namespace
+
+	log := r.log.WithValues("deploymentName", deploy.Name)
+	log.Info("Reconciling controller plugin deployment")
+
+	opResult, err := ctrlutil.CreateOrUpdate(r.ctx, r.Client, deploy, func() error {
+		r.log.Info("Controller plugin deployment reconciled successfully")
```

Comment:
may need to move this log to the end of mutation?

Reply:
This log does not exist anymore. Please confirm this can be resolved

Reply:
resolve.